File: kotlin/spring-data/spring-data-jpa/src/main/kotlin/com/example/UserService.kt
package com.example

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class UserService(private val userRepository: UserRepository) {

    fun findAll(): List<User> = userRepository.findAll()

    fun findById(id: Int): User = userRepository.getById(id)

    fun create(user: User): User = userRepository.save(user)

    fun update(user: User): User {
        val current = userRepository.getById(user.id)
        current.name = user.name
        current.age = user.age
        return userRepository.save(current)
    }

    fun delete(id: Int) {
        val user = userRepository.getById(id)
        userRepository.deleteById(user.id)
    }
}

File: scala/src/main/scala/com/example/UserService.scala
package com.example

import jakarta.persistence.EntityNotFoundException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

import scala.jdk.CollectionConverters.*
import scala.jdk.javaapi.OptionConverters.toScala

@Service
@Transactional
class UserService(val userRepository: UserRepository) {

  def findAll(): Iterable[User] = userRepository.findAll().asScala

  def findById(id: Int): Option[User] = toScala(userRepository.findById(id))

  def create(user: User): User = userRepository.save(user)

  def update(user: User): User = {
    val current = findById(user.id).getOrElse(throw new EntityNotFoundException())
    val updated = current.copy(name = user.name, age = user.age)
    userRepository.save(updated)
  }

  def delete(id: Int): Unit = userRepository.deleteById(id)
}
